vms/platformvm/state/state.go

```diff
@@ -943,74 +935,149 @@ func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
 	return nil
 }
 
-func (s *state) GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error) {
-	prefixStruct := heightWithSubnet{
-		Height:   height,
-		SubnetID: subnetID,
-	}
-	prefixBytes, err := blocks.GenesisCodec.Marshal(blocks.Version, prefixStruct)
-	if err != nil {
-		return nil, err
-	}
-	prefixStr := string(prefixBytes)
-
-	if weightDiffs, ok := s.validatorWeightDiffsCache.Get(prefixStr); ok {
-		return weightDiffs, nil
-	}
-
-	rawDiffDB := prefixdb.New(prefixBytes, s.validatorWeightDiffsDB)
-	diffDB := linkeddb.NewDefault(rawDiffDB)
-	diffIter := diffDB.NewIterator()
+func (s *state) ApplyValidatorWeightDiffs(
+	validators map[ids.NodeID]*validators.GetValidatorOutput,
+	startHeight uint64,
+	endHeight uint64,
+	subnetID ids.ID,
+) error {
+	diffIter := s.flatValidatorWeightDiffsDB.NewIteratorWithStartAndPrefix(
+		getStartWeightKey(subnetID, startHeight),
+		subnetID[:],
+	)
 	defer diffIter.Release()
 
-	weightDiffs := make(map[ids.NodeID]*ValidatorWeightDiff)
+	prevHeight := startHeight + 1
 	for diffIter.Next() {
-		nodeID, err := ids.ToNodeID(diffIter.Key())
+		_, parsedHeight, nodeID, err := parseWeightKey(diffIter.Key())
 		if err != nil {
-			return nil, err
+			return err
 		}
+		if parsedHeight < endHeight {
+			return diffIter.Error()
+		}
+
+		prevHeight = parsedHeight
 
 		weightDiff := ValidatorWeightDiff{}
 		_, err = blocks.GenesisCodec.Unmarshal(diffIter.Value(), &weightDiff)
 		if err != nil {
-			return nil, err
+			return err
+		}
+
+		if err := applyWeightDiff(validators, nodeID, &weightDiff); err != nil {
+			return err
+		}
+	}
+
+	// TODO: Remove this once it is assumed that all subnet validators have
+	// adopted the new indexing.
+	for height := prevHeight - 1; height >= endHeight; height-- {
```

Comment:
If a user rolls back to a prior version - then this will not work correctly when they roll back forward.

Reply:
I don't see where the dropping comes from. It may cause a "hole" in the diffs sequence for the new index, which may be not covered by the old one, thus putting the node in position to not fully reconstruct the diffs. Is this what you mean?

Reply:
IIUC should we introduce a rebuild phase to make sure the two indexes are in sync?

Reply:
I added a range tracker to keep this up-to-date.